pkg/services/blockfetcher/blockfetcher.go

```diff
@@ -0,0 +1,394 @@
+
+	var err error
+	switch bfs.cfg.Mode {
+	case ModeIndexSearch:
+		err = bfs.fetchData()
+	case ModeOidSearch:
+		err = bfs.fetchDataWithOid()
+	default:
+		bfs.log.Error("NeoFS block fetcher service: invalid mode specified", zap.String("mode", bfs.cfg.Mode))
+		return
+	}
+
+	if err != nil {
+		bfs.log.Error("NeoFS block fetcher service: fetch operation failed", zap.Error(err))
+		bfs.Shutdown()
+		return
+	}
+	bfs.Shutdown()
+}
+
+// Shutdown stops the block fetcher service.
+func (bfs *Service) Shutdown() {
+	if bfs.started.CompareAndSwap(true, false) {
+		bfs.log.Info("shutting down NeoFS block fetcher service")
+		bfs.client.Close()
+		close(bfs.quit)
+		<-bfs.bFetcherExitCh
+		_ = bfs.log.Sync()
+	}
+}
+
+// IsActive returns true if the block fetcher service is running.
+func (bfs *Service) IsActive() bool {
+	return bfs.started.Load()
+}
+
+func (bfs *Service) fetchData() error {
+	startIndex := bfs.chain.BlockHeight()
+	batchSize := uint32(bfs.cfg.BatchSize)
+
+	for {
+		select {
+		case <-bfs.quit:
+			bfs.log.Info("NeoFS block fetcher service: stopping NeoFS data fetching operation.")
+			return nil
+		default:
+			prm := client.PrmObjectSearch{}
+			filters := object.NewSearchFilters()
+			filters.AddFilter(bfs.cfg.BlockAttribute, fmt.Sprintf("%d", startIndex), object.MatchNumGE)
+			filters.AddFilter(bfs.cfg.BlockAttribute, fmt.Sprintf("%d", startIndex+batchSize-1), object.MatchNumLE)
+			prm.SetFilters(filters)
+
+			ctx, cancel := context.WithTimeout(context.Background(), bfs.cfg.Timeout)
+
+			blockOids, err := bfs.search(ctx, prm)
+			cancel()
+			bfs.log.Info(fmt.Sprintf("NeoFS block fetcher service: found %d blocks from index %d to %d", len(blockOids), startIndex, startIndex+batchSize-1))
+			if err != nil {
+				bfs.log.Error(fmt.Sprintf("NeoFS block fetcher service: failed to find block with index %d", startIndex), zap.Error(err))
+				return err
+			}
+
+			if len(blockOids) == 0 {
+				bfs.log.Info(fmt.Sprintf("NeoFS block fetcher service: no block found with index %d, stopping.", startIndex))
+				return errors.New("NeoFS block fetcher service: no block found")
+			}
+			err = bfs.fetchAndProcessBlocks(blockOids)
+			if err != nil {
+				return err
+			}
+			startIndex += batchSize
+		}
+	}
+}
+
+func (bfs *Service) fetchAndProcessBlocks(blockOids []oid.ID) error {
+	var wg sync.WaitGroup
+
+	for i, oidBlock := range blockOids {
+		wg.Add(1)
+		go func(oidBlock oid.ID, i int) {
+			defer wg.Done()
+
+			ctx, cancel := context.WithTimeout(context.Background(), bfs.cfg.Timeout)
+			defer cancel()
+			rc, err := bfs.get(ctx, oidBlock.String())
+			if err != nil {
+				bfs.log.Error(fmt.Sprintf("NeoFS block fetcher service: failed to fetch block %s", oidBlock.String()), zap.Error(err))
+				return
+			}
+
+			err = bfs.processBlock(rc, i)
+			if err != nil {
+				bfs.log.Error(fmt.Sprintf("NeoFS block fetcher service: failed to process block with OID %s", oidBlock.String()), zap.Error(err))
+			}
```

Comment:
I see these repetative logs. Why not to store the last error in a method-scoped atomic variable and return the last error from `fetchAndProcessBlock`? Without these logs, because the caller will handle this error.